### PR TITLE
SoF - Rewrite of last scenario

### PR DESCRIPTION
**Pull request by doofus-01**
This doesn't change the plot in any significant way, but it should be a more credible way to end things.  There are two concerns I have:
1.  It may be too easy.  Or it may require too much "behind the scenes knowledge".  I don't know how to test that aside from getting other people to play it.
2.  The micro_ai for the special orc didn't work at all.  I don't know if it is because I was using it wrong or there was a conflict with the new filtering*. I've got a work around, but it should be replaced when possible.

*(https://wiki.wesnoth.org/Wesnoth_AI_Framework#The_.5Bcandidate_action.5D_Tag)

### CODE REVIEW

**Comment by mattsc** [comment]
>     2. The micro_ai for the special orc didn't work at all. 

Maybe I am missing something, but it looks like Gathor is on Side 3, and you are adding the Micro AI to Side 4?

**Comment by doofus-01** [comment]
Thanks for checking.  Gathor being on side 3 is part of the work-around, originally he was part of side 4.

**Comment by mattsc** [comment]
Okay, then I can't find anything that looks like a problem.  It looks like it should work.  Do you have a save from a situation when he should be attacking and did not?

**Comment by doofus-01** [comment]
No, but I can try to recreate it.  I can say he was behaving exactly like everyone else; was not the first to attack, retreated when he was wounded.  I checked the AI config with ":inspect" and saw that the micro_ai was there and had what looked like appropriate [args], but there was nothing in the [filter_own] section (if that matters).

**Comment by mattsc** [comment]
Empty [filter_own] is what it should be in this case.  Let me test a couple things, maybe there's a bug in the code when combining the two filters.  I thought I checked all that already, but it's always possible I missed something.

**Comment by mattsc** [comment]
So I did some test and did not find a problem with the Simple Attack MAI.  The attacks in the test scenario work as expected (I added some output to make sure that they were actually executed by the MAI).  Then I added another MAI using a specific unit's id.  I even copied the code directly from your commit, just in case there's a subtle typo that we're both missing.  I gave it to a peasant with 1 HP and id=Gathor.  Without the MAI he retreats.  With the MAI he is the first to commit suicide.  So as far as I can tell, everything is working as it should.

If you could recreate a save, that would be very useful.

File: data/campaigns/Sceptre_of_Fire/scenarios/9_Caverns_of_Flame.cfg [hunk 1]
```diff
@@ -54,53 +56,124 @@
         {FLAG_VARIANT wood-elvish}
     [/side]
 
-    # rather than have the orcs + trolls be mushed together as generic baddies, let's have the trolls acting as guardians of the volcano and the orcs are more transient
+    # work-around until better AI control is found
```
Comment (stevecotton) [inline data/campaigns/Sceptre_of_Fire/scenarios/9_Caverns_of_Flame.cfg]: Please have the comment say what the issue that's being worked around is.

File: data/campaigns/Sceptre_of_Fire/scenarios/9_Caverns_of_Flame.cfg [hunk 2]
```diff
@@ -434,430 +1027,50 @@
...
-            x,y=14,1
-        [/scroll_to]
-        {QUAKE "rumble.ogg"}
-        {QUAKE "rumble.ogg"}
+        [/redraw]
         [message]
-            speaker=Baglur
-            message= _ "We’ve succeeded in stopping the elves from getting the Sceptre, I'd say.  But what do we do now?!"
+            speaker=Gathor
+            message= _ "Ah ha!  I knew it, there are intruders!  And yet our damned bridge is broken, it was built by idiots.  Come, come, bring planks, we've got to get across!"
```
Comment (stevecotton) [inline data/campaigns/Sceptre_of_Fire/scenarios/9_Caverns_of_Flame.cfg]: The double-space should be a single-space.

File: data/campaigns/Sceptre_of_Fire/scenarios/9_Caverns_of_Flame.cfg [hunk 3]
```diff
@@ -434,430 +1027,50 @@
...
-            speaker=Thursagan
-            message= _ "Keep fighting!  It'll take a miracle to get us out of here, but as long as we stand there is still that slim chance!"
-        [/message]
-        {QUAKE "rumble.ogg"}
-        {QUAKE "rumble.ogg"}
-#ifdef __UNUSED__
         [message]
             speaker=Rugnur
-            message= _ "We cannot let any of the elves or their mercenaries escape; they would tell where the Sceptre of Fire is. Make sure they all die!"
+            message= _ "Huh.  Well, let's not stay here too long."
```
Comment (stevecotton) [inline data/campaigns/Sceptre_of_Fire/scenarios/9_Caverns_of_Flame.cfg]: Please use a typographic apostrophe, both here and on the next message.

File: data/campaigns/Sceptre_of_Fire/scenarios/9_Caverns_of_Flame.cfg [hunk 4]
```diff
@@ -172,6 +300,424 @@
...
+            halo=scenery/rune[1,1,1-glow,1,1,1].png:[200*6]
+        [/item]
+        [event]
+            name=last_breath
+            [filter]
+                id=Gathor
+            [/filter]
+            [message]
+                speaker=unit
+                message= _ "Argh, this hurts!  But I've still got a trick up my sleave..."
```
Comment (nemaara) [inline data/campaigns/Sceptre_of_Fire/scenarios/9_Caverns_of_Flame.cfg]: sleave -> sleeve

**Comment by doofus-01** [comment]
@mattsc - I've tried to reproduce it, and at first all orcs were suicidal, so I reloaded to look at it more, and then it started working (only Gathor was suicidal, the others retreated).  I did have to rearrange things, to 'undo' the work-around, it's possible I introduced a new typo or didn't re-include an old one.  Rather than waste any more of your time, I'll just stick with the work-around because it's safer.  I'll bring it up again if I can reproduce with a simpler test-case.

File: data/campaigns/Sceptre_of_Fire/scenarios/9_Caverns_of_Flame.cfg (hunk 1 above)
Comment (doofus-01) [reply data/campaigns/Sceptre_of_Fire/scenarios/9_Caverns_of_Flame.cfg]: added more to the comments

File: data/campaigns/Sceptre_of_Fire/scenarios/9_Caverns_of_Flame.cfg (hunk 2 above)
Comment (doofus-01) [reply data/campaigns/Sceptre_of_Fire/scenarios/9_Caverns_of_Flame.cfg]: fixed

File: data/campaigns/Sceptre_of_Fire/scenarios/9_Caverns_of_Flame.cfg (hunk 4 above)
Comment (doofus-01) [reply data/campaigns/Sceptre_of_Fire/scenarios/9_Caverns_of_Flame.cfg]: fixed

**Comment by mattsc** [comment]
@doofus-01 - okay, thanks, let me know if you find a reproducible test case.

**Comment by Wedge009** [comment]
I suppose this feedback is better made in the forums discussing the game-play, but I'll just drop my remarks here (for 1.16.1 release): in terms of overall story the rewrite works better than I remember the original being (my save files suggest I last played in 1.4). However, I found it very hard to know what to do. I suppose it's not unreasonable to expect the player to eventually kill Toomak the Troll Hero, thereby destroying the second of the runic tablets (nice that the orc leader is stupid enough to destroy the first one for you). But I had trouble with the final one at (29,31). Clearly only airborne units can reach this location but my Krawg had been excessively harassed by those Cave Wyrmlets so I needed to give him rest in a village, thus my freebie Gryphon Rider (from the start of Hills of the Shorbear Clan) was the first one to reach it. When nothing happened, I assumed the tablet was just a red herring.

Maybe I'll break this into an enhancement request.

**Pull request merged**